crates/typst/src/geom/gradient.rs

```diff
@@ -0,0 +1,1089 @@
+    /// parameter, which is set to 20 by default.
+    ///
+    /// ```example
+    /// #rect(width: 100pt, height: 20pt, fill: gradient.linear(..gradient.cividis(10)))
+    /// ````
+    ///
+    /// [cividis]: https://bids.github.io/colormap/
+    #[func]
+    fn cividis(
+        #[default(Spanned::new(20, Span::detached()))] stops: Spanned<i64>,
+    ) -> SourceResult<Value> {
+        fn at(t: f32) -> Rgba {
+            let t = t.clamp(0.0, 1.0);
+            let r = -4.54
+                - t * (35.34
+                    - t * (2381.73 - t * (6402.7 - t * (7024.72 - t * 2710.57))));
+            let g = 32.49
+                + t * (170.73 + t * (52.82 - t * (131.46 - t * (176.58 - t * 67.37))));
+            let b = 81.24
+                + t * (442.36
+                    - t * (2482.43 - t * (6167.24 - t * (6614.94 - t * 2475.67))));
+
+            Rgba::new(r / 255.0, g / 255.0, b / 255.0, 1.0)
+        }
+
+        if stops.v < 2 {
+            bail!(stops.span, "number of stops must be bigger or equal to 2");
+        }
+
+        Ok(Array::from(
+            (0..stops.v)
+                .map(|i| {
+                    let t = i as f64 / (stops.v - 1) as f64;
+                    Stop::new(Color::Rgba(at(t as f32)), t).into_value()
+                })
+                .collect::<EcoVec<_>>(),
+        )
+        .into_value())
+    }
+
+    /// Creates a list of rainbow color stops with the given parameters.
+    ///
+    /// You can control the number of stops in the gradient using the `stops`
+    /// parameter, which is set to 20 by default.
+    ///
+    /// This color map is best used by setting the interpolation color space to
+    /// [HSL]($color.hsl). It should also be noted that this is not a good
+    /// choice for a color scale, as it is not perceptually uniform. This preset
+    /// is more intended for decorative purposes than for data visualization.
+    ///
+    /// ```example
+    /// #rect(width: 100pt, height: 20pt, fill: gradient.linear(..gradient.rainbow(2)))
+    /// ````
+    #[func]
+    fn rainbow(
+        #[default(Spanned::new(20, Span::detached()))] stops: Spanned<i64>,
+    ) -> SourceResult<Array> {
+        if stops.v < 2 {
+            bail!(stops.span, "number of stops must be bigger or equal to 2");
+        }
+
+        Ok((0..stops.v)
+            .map(|i| {
+                let t = i as f32 / (stops.v - 1) as f32;
+                let ts = (t - 0.5).abs();
+
+                Stop::new(
+                    cubehelix_to_rgb(360.0 * t - 100.0, 1.5 - 1.5 * ts, 0.8 - 0.8 * ts),
+                    t as f64,
+                )
+                .into_value()
+            })
+            .collect())
+    }
+}
+
+impl Gradient {
+    /// Samples the gradient at a given position, in the given container.
+    /// Handles the aspect ratio and angle directly.
+    pub fn sample_at(&self, (x, y): (f32, f32), (width, height): (f32, f32)) -> Color {
```

Comment:
wouldn't it make more sense for this to take normalize coordinates in the first place?

Reply:
No because of the aspect ratio compensation, or it would just take the aspect ratio as a separate param which is basically the same